### PR TITLE
feat: allow :map commands to create maps with description

### PR DESCRIPTION
**Pull request by shadmansaleh**
Now `:map` commands can set map descriptions too . For example
```vim
:map <desc=Save all files and quit nvim> <leader>q <Cmd>wqa<CR>
```

### CODE REVIEW

**Comment by zeertzjq** [comment]
Do you want to deal with a description that contains a `>` character?

**Comment by shadmansaleh** [comment]
> Do you want to deal with a description that contains a > character?

 We could allow `>` to be escaped.
`\>` or `>>` ?

File: runtime/doc/map.txt [hunk 1]
```diff
@@ -502,10 +509,7 @@ Note: When using mappings for Visual mode, you can use the "'<" mark, which
 is the start of the last selected Visual area in the current buffer |'<|.
 
 The |:filter| command can be used to select what mappings to list.  The
-pattern is matched against the {lhs} and {rhs} in the raw form.  If a
-description was added using |nvim_set_keymap()| or |nvim_buf_set_keymap()|
-then the pattern is also matched against it.
-
+pattern is matched against the {lhs}, {rhs} and {desc} in the raw form.
```
Comment (zeertzjq) [inline runtime/doc/map.txt:512]:
```suggestion
pattern is matched against the {lhs}, {rhs} and <desc> in the raw form. 
```

**Comment by zeertzjq** [comment]
> Is there any function the neovim that handles escaping like this ?

In `menu.c` there is a function `menu_unescape_name()` that does something similar.

**Comment by shadmansaleh** [comment]
@zeertzjq escaping `>` with `\` added.

File: runtime/doc/map.txt [hunk 2]
```diff
@@ -199,6 +199,18 @@ Prompts will still be given, e.g., for inputdialog().
 Using "<silent>" for an abbreviation is possible, but will cause redrawing of
 the command line to fail.
 
+						*:map-<desc>* *:map-desc*
+"<desc>" can be used provide a description for keymaps. Example: >
```
Comment (zeertzjq) [inline runtime/doc/map.txt:203]:
```suggestion
"<desc>" can be used provide a description for mappings. Example: >
```

File: runtime/doc/map.txt [hunk 3]
```diff
@@ -199,6 +199,18 @@ Prompts will still be given, e.g., for inputdialog().
...
+	:map <desc=Save all files and quit nvim> <leader>q <Cmd>wqa<CR>
+Text between `<desc=` & `>` will be used as map description.
```
Comment (zeertzjq) [inline runtime/doc/map.txt:205]:
```suggestion
Text between `<desc=` & `>` will be used as description.
```

File: runtime/doc/map.txt [hunk 4]
```diff
@@ -199,6 +199,18 @@ Prompts will still be given, e.g., for inputdialog().
...
+`>` can be used inside description by escaping it with `\`. Because `\`
+is the escape charecter it also needs to be escaped. A lone `\` is just
+ignored. So >
+	:map <desc=escape \> with \\> abc cba
+Is a valid way to set description and the description becomes "escape > with \".
+If a description is given during map creation it'll shown while listing
+ key maps.
```
Comment (zeertzjq) [inline runtime/doc/map.txt:212]:
```suggestion
mappings.
```

File: runtime/doc/map.txt [hunk 5]
```diff
@@ -199,6 +199,18 @@ Prompts will still be given, e.g., for inputdialog().
...
+"<desc>" can be used provide a description for mappings. Example: >
+	:map <desc=Save all files and quit nvim> <leader>q <Cmd>wqa<CR>
+Text between `<desc=` & `>` will be used as description.
+`>` can be used inside description by escaping it with `\`. Because `\`
+is the escape charecter it also needs to be escaped. A lone `\` is just
+ignored. So >
+	:map <desc=escape \> with \\> abc cba
+Is a valid way to set description and the description becomes "escape > with \".
+If a description is given during map creation it'll shown while listing
+ mappings.
```
Comment (zeertzjq) [inline runtime/doc/map.txt:212]: Extra space
```suggestion
mappings.
```

File: runtime/doc/api.txt [hunk 6]
```diff
@@ -1584,11 +1584,10 @@ nvim_set_keymap({mode}, {lhs}, {rhs}, {*opts})             *nvim_set_keymap()*
...
                             |:map-arguments| as keys excluding |<buffer>| but
-                            including |noremap| and "desc". "desc" can be used
-                            to give a description to keymap. When called from
-                            Lua, also accepts a "callback" key that takes a
-                            Lua function to call when the mapping is executed.
-                            Values are Booleans. Unknown key is an error.
+                            including |noremap|. When called from Lua, also
+                            accepts a "callback" key that takes a Lua function
+                            to call when the mapping is executed. Values are
+                            Booleans. Unknown key is an error.
```
Comment (zeertzjq) [inline runtime/doc/api.txt:1590]:
```suggestion
                            Booleans except for "desc". Unknown key is an error.
```

File: runtime/doc/map.txt [hunk 7]
```diff
@@ -502,10 +514,7 @@ Note: When using mappings for Visual mode, you can use the "'<" mark, which
 is the start of the last selected Visual area in the current buffer |'<|.
 
 The |:filter| command can be used to select what mappings to list.  The
-pattern is matched against the {lhs} and {rhs} in the raw form.  If a
-description was added using |nvim_set_keymap()| or |nvim_buf_set_keymap()|
-then the pattern is also matched against it.
-
```
Comment (zeertzjq) [inline runtime/doc/map.txt:508]: Keeping this blank line looks more consistent.

File: src/nvim/api/vim.c [hunk 8]
```diff
@@ -1583,8 +1583,7 @@ ArrayOf(Dictionary) nvim_get_keymap(uint64_t channel_id, String mode)
 /// @param  lhs   Left-hand-side |{lhs}| of the mapping.
 /// @param  rhs   Right-hand-side |{rhs}| of the mapping.
 /// @param  opts  Optional parameters map. Accepts all |:map-arguments|
-///               as keys excluding |<buffer>| but including |noremap| and "desc".
-///               "desc" can be used to give a description to keymap.
+///               as keys excluding |<buffer>| but including |noremap|.
 ///               When called from Lua, also accepts a "callback" key that takes
 ///               a Lua function to call when the mapping is executed.
 ///               Values are Booleans. Unknown key is an error.
```
Comment (zeertzjq) [inline src/nvim/api/vim.c:1589]:
```suggestion
///               Values are Booleans except for "desc". Unknown key is an error.
```

File: src/nvim/getchar.c [hunk 9]
```diff
@@ -2757,6 +2757,27 @@ int str_to_mapargs(const char_u *strargs, bool is_unmap, MapArguments *mapargs)
...
     }
+
+    if (STRNCMP(to_parse, "<desc=", 6) == 0) {
+      const char_u *p = to_parse + 6;
+      garray_T ga = GA_INIT(sizeof(char), 20);
+      while (*p != '>' && *p != NUL) {
+        if (*p == '\\') {
+          p++;  // skip escape char
+        }
+        ga_append(&ga, (char) *p);
```
Comment (zeertzjq) [inline src/nvim/getchar.c:2768]:
```suggestion
        ga_append(&ga, (char)(*p));
```

Comment (zeertzjq) [reply src/nvim/getchar.c:2768]: You dropped this in a rebase.

Comment (shadmansaleh) [reply src/nvim/getchar.c:2768]: Oops . I forgot to pull before rebase :P

**Comment by zeertzjq** [comment]
I added a commit to reorganize tests.

File: src/nvim/getchar.c (hunk 9 above)
Comment (shadmansaleh) [reply src/nvim/getchar.c]: added.

**Comment by shadmansaleh** [comment]
Should we handle `\n`, `\t` in descriptions ?

**Comment by zeertzjq** [comment]
I just found a typo when reorganizing the tests, so I force-pushed again.

**Comment by shadmansaleh** [comment]
> I just found a typo when reorganizing the tests, so I force-pushed again.

no problem you can push . I'll make sure to pull before pushing changes.

Also https://github.com/neovim/neovim/pull/17426#issuecomment-1041470639 ☝️?

**Comment by zeertzjq** [comment]
There are many escape sequences, and it can be hard to decide what to support and what not to support. I think it is better to wait for someone else's review on this.

**Comment by zeertzjq** [comment]
There are many escape sequences, and it can be hard to decide what to support and what not to support. I think it is better to wait for someone else's review on this.

**Comment by shadmansaleh** [comment]
> There are many escape sequences, and it can be hard to decide what to support and what not to support. I think it is better to wait for someone else's review on this.

Sure . But I think only \n & \t will actually be useful here for formating. There's no point in escaping vim keycodes like \\\<cmd> in description . Stuff like \a, \b .... doesn't make much sense either. Maybe we could support \r & \v too .

**Comment by jamessan** [comment]
What's the reason for adding this to the `:map` command, instead of just using the API function?

**Comment by shadmansaleh** [comment]
> What's the reason for adding this to the :map command, instead of just using the API function?

Why not ? Since ex commands can also have description allowing ex interface to set them too seems like a good idea .

**Comment by jamessan** [comment]
> Why not ? Since ex commands can also have description allowing ex interface to set them too seems like a good idea .

We should ask "Why?" when adding functionality, not "Why not?".  Adding this to `:map` is awkward because a new way to pass a chunk of free-form text has to be created.  The attributes of `:map` are odd enough as is.

The API already provides an easy way to do this, since it's just dealing with strings.  No need to fight against the limitations of Vim's command parsing.

I didn't follow the PR adding the support for descriptions in the first place, so if it was suggested there to do this, then that's fine.  It just feels like trying to shoe horn functionality into an odd interface for the sake of parity, when we have a perfectly good interface to do this already.

**Comment by zeertzjq** [comment]
Actually I also had the same thought when I began testing this.

**Comment by shadmansaleh** [comment]
Ok . We can leave viml syntax upto vim.

**Pull request closed without merging**